x/gamm/pool-models/balancer/pool.go

```diff
@@ -714,20 +727,22 @@ func (p *Pool) CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee s
 	// * numShares is how many shares are perfectly matched.
 	// * remainingTokensIn is how many coins we have left to join, that have not already been used.
 	// if remaining coins is empty, logic is done (we joined all tokensIn)
-	numShares, remainingTokensIn, err := cfmm_common.MaximalExactRatioJoin(p, sdk.Context{}, tokensIn)
+	numShares, tokensJoined, remainingTokensIn, err := p.CalcJoinPoolNoSwapShares(ctx, tokensIn, swapFee)
 	if err != nil {
 		return sdk.ZeroInt(), sdk.NewCoins(), err
 	}
-	if remainingTokensIn.Empty() {
-		tokensJoined = tokensIn
+
+	// safely ends the calculation if all input tokens are successfully LP'd
+	if tokensJoined.IsAnyGT(tokensIn) {
+		return sdk.ZeroInt(), sdk.NewCoins(), errors.New("an error has occurred, more coins joined than tokens passed in")
+	} else if remainingTokensIn.Empty() {
 		return numShares, tokensJoined, nil
 	}
 
 	// 4) Still more coins to join, so we update the effective pool state here to account for
 	// join that just happened.
 	// * We add the joined coins to our "current pool liquidity" object (poolAssetsByDenom)
 	// * We increment a variable for our "newTotalShares" to add in the shares that've been added.
-	tokensJoined = tokensIn.Sub(remainingTokensIn)
```

Comment:
Context: `CalcJoinPoolNoSwapShares` includes this calculation and returns the resulting `tokensJoined` above